// dolober/dist/js/dropdown.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
	typeof define === 'function' && define.amd ? define(factory) :
	(global.Dropdown = factory());
}(this, (function () { 'use strict';

/*
 * utilities functions
 */
function select(element) {
  if (typeof element === 'string') {
    return document.querySelector(element);
  }
  return element;
}

/**
 * modal class
 */
class Dropdown {
  constructor(selector) {
    this.el = select(selector);
    this.init();
  }

  /**
   * automatically init all modals in the page
   */
  static create() {
    const dropdowns = Array.from(document.querySelectorAll('.dropdown'));
    dropdowns.forEach((dropdown) => {
      new Dropdown(dropdown);
    });
  }

 /**
  * add event listeners to the modal elements
  * @param {HTMLElement} button
  * @param {HTMLElement} element
  */
  init() {
    this.button = this.el.querySelector('.button');
    this.menu = this.el.querySelector('.dropdown-menu');
    this.animating = false;

    this.initEvents();
  }

  initEvents() {
    this.button.addEventListener('click', this.toggleVisibility.bind(this), false);
  }
  /**
   * show hidden modal
   */
  toggleVisibility() {
    this.menu.classList.toggle('is-active');
  }
}

return Dropdown;

})));
