// dolober/src/js/switcher.js
/**
 * event handle when mousemove
 * @param  {object} switcher
 */
function drag(switcher) {
  let dragValue = 0;
  if (event.type === 'mousemove' && event.buttons === 1) {
    dragValue = (event.clientX - switcher.currentX);
  }
  if (event.type === 'touchmove') {
    dragValue = (event.touches[0].clientX - switcher.currentX);
  }
  // TODO: improve this algorithm
  switcher.position = switcher.handle.style.left;
  const moveRange = (switcher.input.checked ? -1 : 1) * (switcher.position + dragValue);
  if (moveRange >= 0 && moveRange <= 30) {
    if (dragValue < 0) {
      switcher.position += dragValue;
    }
    if (Math.abs(dragValue) < 10) {
      switcher.handle.style.width = `${26 + Math.abs(dragValue)}px`;
    }
    if (dragValue > 10) {
      switcher.position += dragValue - 10;
    }
    switcher.handle.style.transform = `translateX(${switcher.position}px)`;
  }
}

/**
 * event handler for mouseup
 * @param {object} switcher
 */
function relase(switcher) {
  event.preventDefault();
  switcher.handle.style.transform = '';
  switcher.handle.style.width = '';
  switcher.handle.classList.remove('is-dragging');
  if (Math.abs(switcher.position) >= 15 || Math.abs(switcher.position) === 0) {
    switcher.input.checked = !switcher.input.checked;
  }
  document.removeEventListener('mousemove', switcher.tempDrag);
  document.removeEventListener('mouseup', switcher.tempRelase);
  document.removeEventListener('touchmove', switcher.tempDrag);
  document.removeEventListener('touchend', switcher.tempRelase);
}

/**
 * init the switchers darging
 */
export default function switcherInit() {
  const switchers = Array.from(document.querySelectorAll('.switcher'));
  switchers.forEach((item) => {
    const switcher = {
      item:       item,
      body:       item.querySelector('.switcher-body'),
      handle:     item.querySelector('.switcher-handle'),
      input:      item.querySelector('.switcher-input'),
      currentX:   0,
      tempDrag:   '',
      tempRelase: '',
      position:   '',
    }

    switcher.body.addEventListener('touchstart', () => {
      event.preventDefault();
      switcher.handle.classList.add('is-dragging');
      switcher.currentX = event.touches[0].clientX;
      document.addEventListener('touchmove', switcher.tempDrag = () => drag(switcher), false);
      document.addEventListener('touchend', switcher.tempRelase = () => relase(switcher), false);
    }, false);

    switcher.body.addEventListener('mousedown', () => {
      event.preventDefault();
      switcher.handle.classList.add('is-dragging');
      switcher.currentX = event.clientX;
      document.addEventListener('mousemove', switcher.tempDrag = () => drag(switcher), false);
      document.addEventListener('mouseup', switcher.tempRelase = () => relase(switcher), false);
    }, false);

    switcher.body.addEventListener('click', () => {
      event.preventDefault();
    }, false);
  });
}

switcherInit();
